File: cpp/example6.cpp
#include <iostream>
#include <cstdlib>
#include "exevalator.hpp"
#include "exevalator.cpp"

/**
 * An example to compute the value of the inputted expression f(x) at the inputted x.
 */
int main() {

    // Get the expression from the standard-input
    std::string expression;
    std::cout << "" << std::endl;
    std::cout << "This program computes the value of f(x) at x." << std::endl;
    std::cout << "" << std::endl;
    std::cout << "f(x) = ?               (default: 3*x*x + 2*x + 1)" << std::endl;
    std::getline(std::cin, expression);
    if (expression.empty()) {
        expression = "3*x*x + 2*x + 1";
    }

    // Get the value of the x from the standard-input
    std::string x_str;
    std::cout << "x = ?                  (default: 1)" << std::endl;
    std::getline(std::cin, x_str);
    if (x_str.empty()) {
        x_str = "1";
    }
    double x_value = std::stod(x_str);

    // Create an instance of Exevalator Engine
    exevalator::Exevalator exevalator;

    // Set the value of x
    exevalator.declare_variable("x");
    exevalator.write_variable("x", x_value);

    try {

        // Evaluate the value of an expression
        double result = exevalator.eval(expression);

        // Display the result
        std::cout << "----------" << std::endl;
        std::cout << "f(x)   = " << expression << std::endl;
        std::cout << "x      = " << x_value << std::endl;
        std::cout << "result = " << result << std::endl;

    } catch (exevalator::ExevalatorException &e) {
        std::cout << "Error occurred: " << e.what() << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}

File: rust/example6.rs
mod exevalator;
use exevalator::Exevalator;
use std::io;

/// An example to compute the value of the inputted expression f(x) at the inputted x.
fn main() {

    println!("");
    println!("This program computes the value of f(x) at x.");
    println!("");

    // Get the expression from the standard-input
    println!("f(x) = ?               (example: 3*x+1)");
    let mut expression_string: String = String::new();
    io::stdin().read_line(&mut expression_string).expect("I/O error occurred for receiving the expression.");
    let expression_str = expression_string.trim();

    // Get the value of x from the standard-input
    println!("x = ?                  (example: 2)");
    let mut x_string: String = String::new();
    io::stdin().read_line(&mut x_string).expect("I/O error occurred for receiving the x value.");
    let x_value: f64 = x_string.trim().parse::<f64>().expect("Invalid x value.");

    // Create an instance of Exevalator Engine
    let mut exevalator = Exevalator::new();

    // Set the value of x
    exevalator.declare_variable("x").expect("Failed to declare the variable x.");
    exevalator.write_variable("x", x_value).expect("Failed to set the value of the variable x.");

    // Evaluate the value of an expression
    let result: f64 = match exevalator.eval(expression_str) {
        Ok(eval_value) => eval_value,
        Err(eval_error) => panic!("{}", eval_error),
    };

    // Display the result
    println!("----------");
    println!("f(x)   = {}", expression_str);
    println!("x      = {}", x_value);
    println!("result = {}", result);
}
